### ISSUE TITLE
SIGSEGV: Illegal storage access. (Attempt to read from nil?) in generic proc

### ISSUE THREAD

**Issue by danilaml** [nim-lang/Nim#6073]
This code fails to compile for me on Nim (`Nim Compiler Version 0.17.0 (2017-05-22) [Linux: amd64]
`) on Ubuntu 16.04 with `SIGSEGV: Illegal storage access. (Attempt to read from nil?)` error.
```nim
import system
 
type Bar[T] = ref object
  value: T
 
type types = int32|int64 # if I change this to just int32 or int64 it works (compiles)

# if I replace Bar everywhere with seq it also compiles fine 
proc Foo[T: Bar[types]](): T =
  when T is Bar: nil
 
discard Foo[Bar[int32]]()
```

However it compiles just fine on `0.16.0` (tested on ideone), so it looks like a regression.

**Comment by TiberiumN** [comment]
Debug output from compiler on devel:
```bash
Traceback (most recent call last)                                                                           
nim.nim(121)             nim
nim.nim(77)              handleCmdLine
main.nim(163)            mainCommand
main.nim(74)             commandCompileToC
modules.nim(240)         compileProject
modules.nim(180)         compileModule
passes.nim(215)          processModule
passes.nim(135)          processTopLevelStmt
sem.nim(550)             myProcess
sem.nim(522)             semStmtAndGenerateGenerics
semstmts.nim(1837)       semStmt
semexprs.nim(820)        semExprNoType
semexprs.nim(2346)       semExpr
semstmts.nim(19)         semDiscard
semexprs.nim(44)         semExprWithType
semexprs.nim(2260)       semExpr
semexprs.nim(731)        semIndirectOp
semexprs.nim(2273)       semExpr
semexprs.nim(1295)       semArrayAccess
semexprs.nim(1271)       semSubscript
semcall.nim(461)         explicitGenericInstantiation
sigmatch.nim(326)        explicitGenericSym
sigmatch.nim(1550)       typeRelImpl
sigmatch.nim(1375)       typeRelImpl
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
```

**Comment by carlocapocasa** [comment]
Just attempted this on 0.17.2 and arch linux, SIGSEGV: Illegal storage access. (Attempt to read from nil?)
occurs again.

**Comment by Yardanico** [comment]
@carlocapocasa yeah, it was fixed after 0.17.2 release

**Comment by carlocapocasa** [comment]
ok thx, sorry
